README.md
# strider-gitlab

[![Greenkeeper badge](https://badges.greenkeeper.io/Strider-CD/strider-gitlab.svg)](https://greenkeeper.io/)

> Note: This currently only works with Gitlab API v3

This plugin provides seamless integration between [Strider](https://github.com/Strider-CD/strider) and [GitLab](https://github.com/gitlabhq/gitlabhq). It enables you to add GitLab repositories to Strider and it hooks up necessary webhooks in order to automatically test on commits.

[![Build Status](https://travis-ci.org/Strider-CD/strider-gitlab.svg)](https://travis-ci.org/Strider-CD/strider-gitlab)

## Set Up

1. Install the plugin (can be done either from the Admin/Plugins section of the GUI or by running `bin/strider install gitlab`)
2. Click on your e-mail address in the top right corner of the GUI, then click "Account"
3. Click "GitLab" at the left of your screen
4. Add your information (the private token can be found in your account settings on GitLab)
5. Navigate to "Projects" in the Strider GUI. You should now see a tab with the name of your GitLab account

## Contributors

- Martin Ericson (http://www.devbox.com)
- [nodefourtytwo/strider-gitlab](https://github.com/nodefourtytwo/strider-gitlab)
- [jonlil](https://github.com/jonlil)
- [edy](https://github.com/edy)
- [_others_](https://github.com/Strider-CD/strider-gitlab/graphs/contributors)

## License

The MIT License (MIT)

Copyright (c) 2014 Martin Ericson

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [2.0.0](https://github.com/Strider-CD/strider-gitlab/compare/v1.4.0...v2.0.0) (2020-05-02)


### ⚠ BREAKING CHANGES

* Uses GitLab v4 API

### Bug Fixes

* convert to v4 API ([9bba287](https://github.com/Strider-CD/strider-gitlab/commit/9bba287c66f22fdaeffc55a8f346eee5047f8628))

## [1.4.0](https://github.com/Strider-CD/strider-gitlab/compare/v1.3.1...v1.4.0) (2020-05-02)


### Features

* add option to limit gitlab results to projects you are a member of ([73b11df](https://github.com/Strider-CD/strider-gitlab/commit/73b11df43ea8a4d8491d02d9829f7bde6ca3a7ee))

### [1.3.1](https://github.com/Strider-CD/strider-gitlab/compare/v1.3.0...v1.3.1) (2020-05-02)


### Bug Fixes

* actually fix the page issue ([a83536c](https://github.com/Strider-CD/strider-gitlab/commit/a83536c6c60155880658cf20c40254d36cff63fd))

## [1.3.0](https://github.com/Strider-CD/strider-gitlab/compare/v1.2.2...v1.3.0) (2020-05-02)


### Features

* Tag push support ([#110](https://github.com/Strider-CD/strider-gitlab/issues/110)) ([f7f8d61](https://github.com/Strider-CD/strider-gitlab/commit/f7f8d618fe0cc26305e4adc297dde9d01f0d29fc))


### Bug Fixes

* default to 1 page if unknown ([433b279](https://github.com/Strider-CD/strider-gitlab/commit/433b279725e6240d2d161cfce5e077446a855731))
* **package:** update async to version 2.6.0 ([#111](https://github.com/Strider-CD/strider-gitlab/issues/111)) ([7c3da7a](https://github.com/Strider-CD/strider-gitlab/commit/7c3da7a8bfd1acc2d6ea439bd0d9b769a68faf3c)), closes [#92](https://github.com/Strider-CD/strider-gitlab/issues/92)
* **package:** update superagent to version 3.6.0 ([#99](https://github.com/Strider-CD/strider-gitlab/issues/99)) ([8899ff8](https://github.com/Strider-CD/strider-gitlab/commit/8899ff866e751afb1ad900a14ca6eb897a6682a0))
* webhook endpoint ([4268b2c](https://github.com/Strider-CD/strider-gitlab/commit/4268b2cc31311276a01cf8fd9e1f0050ecca13a7)), closes [#54](https://github.com/Strider-CD/strider-gitlab/issues/54)

## 1.0.4

Bugfixes:

  - fixed an issue if group name is changed in gitlab

## 1.0.3

Features:

  - #4

Bugfixes:

  - added gravatar requirement to package.json

## 1.0.2

Bugfixes:

  - #2


test/test_api.js
'use strict';

/*

 Used to test the gitlab API wrapper

 nock will simulate a gitlab server running at
 localhost:80, where Strider Tester, a user is
 registered with the name "stridertester", and
 has been registered with api token - zRtVsmeznn7ySatTrnrp
 stridertester is an "owner" of a group named "testunion"
 and has admin access to three projects -
 testunion / unionproject1
 Strider Tester / pubproject1
 Strider Tester / privproject1
 */


const expect = require('expect.js');
const api = require('../lib/api');
let util = require('util');
let debug = require('debug')('strider-gitlab:test:api');
const nock = require('nock');

const correctConfig = {
  api_key: 'zRtVsmeznn7ySatTrnrp',
  api_url: 'http://localhost:80/api/v3'
};

const wrongCredentialsConfig = {
  api_key: 'zRtVsmeznn7ySatTrnra',
  api_url: 'http://localhost:80/api/v3'
};

const invalidServerNameConfig = {
  api_key: 'zRtVsmeznn7ySatTrnrp',
  api_url: 'http://localghost:80/api/v3'
};

const configWithoutApiUrl = {api_key: 'zRtVsmeznn7ySatTrnrp'};

const configWithoutApiKey = {
  api_url: 'http://localhost:80/api/v3'
};

const correctDeployKey = "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDAMoSHhKfeE3/oXanAQEZO0Sq20SMjvjmJlTy+CaGz/1uk+glLXi9u2RKtfPRZDceAgyEtRUpqya9Uo1v9bjkIckGLhQwXdSo2G6O3QuzpE3gc6AXTDPQ0ZkkXbSdU9VGL1Zzr+maBnvfwK6IlsNz3fLa4lNV7vz1LaGCg9D1jP+nufZjuDiCAno7D607oG1iHQ3x/BqzphUATav3DFQFT2FBmmittQT0l0mMJ4XsQCQXkwNbDjkLYNon8FYPm9U3AOlzicOGteebt5mhsQtfl9+lL99B8+fk8b24pEEbOxZ4l0HcwMI1R5OLoTzPwSvVw+bp3YPhH2IzfFwK5NUk7 stridertester/privproject1-stridertester@gmail.com\n";

describe('gitlab api', function() {
  before('Setup the mock gitlab server', function setupNock() {
    nock.cleanAll();
    require('./mocks/gitlab_get.js')();
    require('./mocks/gitlab_add_key.js')();
  });

  after('Tear down mock Gitlab server', function tearDownNock() {
    nock.cleanAll();
  });

  //--------------------------------------------------------------------------------------
  describe('get', function() {
    it('should get a list of projects', function(done) {
      api.get(correctConfig, 'projects', function(err, body, res) {
        expect(err).to.not.be.ok();
        expect(body).to.be.an(Array);
        done();
      });
    });

    it('should return an error if a non existent path is asked for', function(done) {

      const wrongPath = 'nonexistentpath';

      api.get(correctConfig, wrongPath, function(err, body, res) {
        expect(err).to.be.ok();
        expect(err).to.be.an(Error);
        expect(err.status).to.eql('404');
        done();
      });
    });

    it('should return a 401 error if incorrect credentials are specified', function(done) {
      api.get(wrongCredentialsConfig, 'projects', function(err, body, res) {
        expect(err).to.be.ok();
        expect(err).to.be.an(Error);
        expect(err.status).to.eql('401');
        done();
      });
    });

    it('should return an error if the specified gitlab server IP cannot be resolved from the name', function(done) {
      api.get(invalidServerNameConfig, 'projects', function(err, body, res) {
        expect(err).to.be.ok();
        expect(err).to.be.an(Error);
        done();
      });
    });

    it('should return an error if the config passed in to it does not have a Gitlab API url', function(done) {
      const config = {api_key: 'zRtVsmeznn7ySatTrnrp'};
      api.get(config, 'projects', function(err, body, res) {
        expect(err).to.be.ok();
        expect(err).to.be.an(Error);
        done();
      });
    });

    it('should invoke our callback with the first parameter as undefined if the Gitlab server responds with a falsy value for res.body');
  });

  //--------------------------------------------------------------------------------------
  describe('parseRepo', function() {
    it('should correctly parse repo information as received from gitlab server into a repo object');
    it('should throw an error if it gets an empty parameter as repo ?');
    it('should throw an error if repo.id is absent ?');
    it('should throw an error if repo.path_with_namespace is absent ?');
    it('should throw an error if repo.web_url is absent ?');
    it('should throw an error if repo.public is absent ?');
    it('should throw an error if repo.ssh_url_to_repo is absent ?');
    it('should throw an error if repo.owner is absent ?');
    it('should throw an error if repo.web_url is absent ?');

    /*
     return {
     83     id: repo.id,
     84     name: repo.path_with_namespace,
     85     display_name: repo.path_with_namespace,
     86     display_url: repo.web_url,
     87     group: repo.namespace.path,
     88     private: !repo.public,
     89     config: {
     90       auth: {type: 'ssh'},
     91       scm: 'git',
     92       url: repo.ssh_url_to_repo,
     93       owner: repo.owner,
     94       repo: repo.web_url,
     95       pull_requests: 'none',
     96       whitelist: []
     97     }
     */
  });

  //--------------------------------------------------------------------------------------
  describe('addDeployKey', function() {
    it('should return an error if any of its expected arguments are absent');

    it('should invoke our callback with err as null and the second parameter as true, when given correct parameters', function(done) {
      api.addDeployKey(correctConfig, 5, 'strider-stridertester/privproject1', correctDeployKey, function(err, secondParam) {
        expect(err).to.not.be.ok();
        expect(secondParam).to.be.ok();
        done();
      });
    });

    it('should give an error if invalid data is sent as an ssh key', function(done) {
      api.addDeployKey(correctConfig, 5, 'strider-stridertester/privproject1', "invalid-key", function(err, secondParam) {
        //debug("Err is: " + util.inspect(err, false, 10, true));
        expect(err).to.be.ok();
        done();
      });
    });

    it('should give an error if incorrect credentials are passed to it', function(done) {
      api.addDeployKey(wrongCredentialsConfig, 5, 'strider-stridertester/privproject1', correctDeployKey, function(err, secondParam) {
        //debug("Err is: " + util.inspect(err, false, 10, true));
        expect(err).to.be.ok();
        done();
      });
    });

    it('should return an error if the specified gitlab server IP cannot be resolved from the name', function(done) {
      api.addDeployKey(invalidServerNameConfig, 5, 'strider-stridertester/privproject1', correctDeployKey, function(err, secondParam) {
        expect(err).to.be.ok();
        expect(err).to.be.an(Error);
        done();
      });
    });

    it('should return an error if the config passed in to it does not have a Gitlab API url', function(done) {
      const config = {api_key: 'zRtVsmeznn7ySatTrnrp'};
      api.addDeployKey(config, 5, 'strider-stridertester/privproject1', correctDeployKey, function(err, secondParam) {
        expect(err).to.be.ok();
        expect(err).to.be.an(Error);
        done();
      });
    });

    it('should give an error if invalid repo id is passed to it', function(done) {
      api.addDeployKey(correctConfig, "wrong repo id", 'strider-stridertester/privproject1', correctDeployKey, function(err, secondParam) {
        //debug("Err is: " + util.inspect(err, false, 10, true));
        expect(err).to.be.ok();
        done();
      });
    });
  });

  //--------------------------------------------------------------------------------------
  describe('removeDeployKey - when key is registered with the server', function() {
    before('Setup the mock gitlab server', function setupNock() {
      nock.cleanAll();
      require('./mocks/gitlab_delete_project.js')();
    });

    after('Tear down mock Gitlab server', function tearDownNock() {
      nock.cleanAll();
    });

    it('should return an error if any of its expected arguments are absent');

    it('should delete a key and invoke our callback with err as null and wasDeleted as true when given correct parameters', function(done) {
      api.removeDeployKey(correctConfig, 5, 'strider-stridertester/privproject1', function(err, wasDeleted) {
        //debug("Err is: " + util.inspect(err, false, 10, true));
        //debug("Err is: " + err);
        //debug("wasDeleted is: " + wasDeleted);
        expect(err).to.not.be.ok();
        expect(wasDeleted).to.be.ok();
        done();
      });
    });

    it('should return an error if wrong credentials are given', function(done) {
      api.removeDeployKey(wrongCredentialsConfig, 5, 'strider-stridertester/privproject1', function(err, wasDeleted) {
        //debug("Err is: " + util.inspect(err, false, 10, true));
        //debug("Err is: " + err);
        expect(err).to.be.ok();
        expect(err).to.be.an(Error);
        done();
      });
    });

    it('should return an error if the config passed in to it does not have a Gitlab API url', function(done) {
      api.removeDeployKey(configWithoutApiUrl, 5, 'strider-stridertester/privproject1', function(err, wasDeleted) {
        //debug("Err is: " + util.inspect(err, false, 10, true));
        //debug("Err is: " + err);
        expect(err).to.be.ok();
        expect(err).to.be.an(Error);
        done();
      });
    });

    it('should return an error if the specified gitlab server IP cannot be resolved from the name', function(done) {
      api.removeDeployKey(invalidServerNameConfig, 5, 'strider-stridertester/privproject1', function(err, wasDeleted) {
        //Currently due to an issue with nock/superagent interaction we get a TypeError thrown instead of ENOTFOUND
        //https://github.com/pgte/nock/issues/211#issuecomment-133636234
        //debug("Err is: " + util.inspect(err, false, 10, true));
        //debug("Err is: " + err.stack);
        expect(err).to.be.ok();
        expect(err).to.be.an(Error);
        done();
      });
    });

    it('should give an error if invalid repo id is passed to it', function(done) {
      api.removeDeployKey(correctConfig, "wrong repo id", 'strider-stridertester/privproject1', function(err, secondParam) {
        //debug("Err is: " + err);
        //debug("Err is: " + util.inspect(err, false, 10, true));
        expect(err).to.be.ok();
        done();
      });
    });
  });

  describe('removeDeployKey - when key is not registered in the gitlab server', function() {
    before('Setup the mock gitlab server', function setupNock() {
      nock.cleanAll();
      require('./mocks/gitlab_delete_key_when_absent.js')();
    });

    after('Tear down mock Gitlab server', function tearDownNock() {
      nock.cleanAll();
    });

    it('should callback with err as null and wasDeleted as false', function(done) {
      api.removeDeployKey(correctConfig, 5, 'strider-stridertester/privproject1', function(err, wasDeleted) {
        //debug("Err is: " + util.inspect(err, false, 10, true));
        //debug("wasDeleted is: " + wasDeleted);
        expect(err).to.not.be.ok();
        expect(wasDeleted).to.not.be.ok();
        done();
      });
    });
  });

  //--------------------------------------------------------------------------------------
  describe('createHooks', function() {
    //takes parameters config, repo_id, url and callback

    beforeEach('Setup the mock gitlab server for creating hooks', function setupNock() {
      nock.cleanAll();
      require('./mocks/gitlab_create_hooks.js')();
    });

    afterEach('Tear down mock Gitlab server', function tearDownNock() {
      nock.cleanAll();
    });


    it('should return true as the second parameter and err as false if hooks were created successfully', function(done) {
      api.createHooks(correctConfig, '5', 'http://localhost:3000/stridertester/privproject1/api/gitlab/webhook', function(err, couldCreateHooks) {
        //debug("error: " + err + " couldCreateHooks: " + couldCreateHooks);
        expect(err).to.not.be.ok();
        expect(couldCreateHooks).to.be.ok();
        done();
      });
    });//NOTE: if the webhook has already been created, gitlab creates a new one with an incremented ID

    it('should callback with an error if config does not have an api_url', function(done) {
      api.createHooks(configWithoutApiUrl, '5', 'http://localhost:3000/stridertester/privproject1/api/gitlab/webhook', function(err, couldCreateHooks) {
        expect(err).to.be.ok();
        done();
      });
    });

    it('should callback with an error if config does not have an api_key', function(done) {
      api.createHooks(configWithoutApiKey, '5', 'http://localhost:3000/stridertester/privproject1/api/gitlab/webhook', function(err, couldCreateHooks) {
        expect(err).to.be.ok();
        done();
      });
    });

    it('should callback with an error if invalid credentials were specified', function(done) {
      api.createHooks(wrongCredentialsConfig, '5', 'http://localhost:3000/stridertester/privproject1/api/gitlab/webhook', function(err, couldCreateHooks) {
        expect(err).to.be.ok();
        done();
      });
    });


    it('should callback with an error if an invalid repo_id was specified', function(done) {
      api.createHooks(correctConfig, 'invalid-repo', 'http://localhost:3000/stridertester/privproject1/api/gitlab/webhook', function(err, couldCreateHooks) {
        expect(err).to.be.ok();
        done();
      });
    });

    it('should callback with a 400 error if the URL parameter is not a string', function(done) {
      api.createHooks(correctConfig, '5', false, function(err, couldCreateHooks) {
        expect(err).to.be.ok();
        done();
      });
    });

    it('should callback with an error if the gitlab server cannot be reached or resolved', function(done) {
      api.createHooks(invalidServerNameConfig, '5', false, function(err, couldCreateHooks) {
        expect(err).to.be.ok();
        done();
      });
    });
  });

  //--------------------------------------------------------------------------------------
  describe('deleteHooks - when hook has been registered', function() {
    before('Setup the mock gitlab server', function setupNock() {
      nock.cleanAll();
      require('./mocks/gitlab_delete_hooks_when_present.js')();
    });

    after('Tear down mock Gitlab server', function tearDownNock() {
      nock.cleanAll();
    });


    it('should call our callback with second param as true and err as null if hooks were successfully deleted', function(done) {
      api.deleteHooks(correctConfig, '5', 'http://localhost:3000/stridertester/privproject1/api/gitlab/webhook', function(err, wasDeleted) {
        expect(err).to.not.be.ok();
        expect(wasDeleted).to.be.ok();
        done();
      });
    });

    it('should callback with an error if config does not have an api_url', function(done) {
      api.deleteHooks(configWithoutApiUrl, '5', 'http://localhost:3000/stridertester/privproject1/api/gitlab/webhook', function(err, wasDeleted) {
        //debug('Error is ' + util.inspect(err, false, null, true));
        expect(err).to.be.ok();
        done();
      });
    });

    it('should callback with an error if config does not have an api_key', function(done) {
      api.deleteHooks(configWithoutApiKey, '5', 'http://localhost:3000/stridertester/privproject1/api/gitlab/webhook', function(err, wasDeleted) {
        expect(err).to.be.ok();
        done();
      });
    });

    it('should callback with an error if invalid credentials were specified', function(done) {
      api.deleteHooks(wrongCredentialsConfig, '5', 'http://localhost:3000/stridertester/privproject1/api/gitlab/webhook', function(err, wasDeleted) {
        expect(err).to.be.ok();
        done();
      });
    });


    it('should callback with an error if an invalid repo_id was specified', function(done) {
      api.deleteHooks(correctConfig, 'invalid-repo', 'http://localhost:3000/stridertester/privproject1/api/gitlab/webhook', function(err, wasDeleted) {
        expect(err).to.be.ok();
        done();
      });
    });


    it('should callback with an error if the gitlab server cannot be reached', function(done) {
      api.deleteHooks(invalidServerNameConfig, '5', 'http://localhost:3000/stridertester/privproject1/api/gitlab/webhook', function(err, wasDeleted) {
        expect(err).to.be.ok();
        done();
      });
    });
  });

  //--------------------------------------------------------------------------------------
  describe('deleteHooks - when hook has not been registered', function() {
    before('Setup the mock gitlab server', function setupNock() {
      nock.cleanAll();
      require('./mocks/gitlab_delete_hooks_when_absent.js')();
    });

    after('Tear down mock Gitlab server', function tearDownNock() {
      nock.cleanAll();
    });

    it('should callback with err as null and wasDeleted as false', function(done) {
      api.deleteHooks(correctConfig, 5, 'strider-stridertester/privproject1', function(err, wasDeleted) {
        //debug("Err is: " + util.inspect(err, false, 10, true));
        //debug("wasDeleted is: " + wasDeleted);
        expect(err).to.not.be.ok();
        expect(wasDeleted).to.not.be.ok();
        done();
      });
    });
  });
});


test/test_worker.js
'use strict';

/*
 Used to test the gitlab worker
 */

const proxyquire = require('proxyquire');
const expect = require('expect.js');
let util = require('util');
let debug = require('debug')('strider-gitlab:test:webapp');

const gitStub = {};

gitStub.fetch = function(dest, config, job, context, done) {
  done();
};

const worker = proxyquire('../lib/worker', {'strider-git/worker': gitStub});

describe('gitlab worker', function() {
  //--------------------------------------------------------------------------------------
  describe('init', function() {
    const dirs = {
      base: '/home/dev/.strider',
      data: '/home/dev/.strider/data/stridertester-privproject1-55de9a6c1cfc0d872d309dbb',
      cache: '/home/dev/.strider/cache/stridertester/privproject1'
    };

    const account = {
      api_url: 'http://localhost:80/api/v3',
      api_key: 'zRtVsmeznn7ySatTrnrp'
    };

    const config = {
      whitelist: [],
      pull_requests: 'none',
      repo: 'http://nodev/stridertester/privproject1',
      owner: {
        avatar_url: 'http://www.gravatar.com/avatar/3f671ed86ed3d21ed3640c7a016b0997?s=40&d=identicon',
        state: 'active',
        id: 3,
        username: 'stridertester',
        name: 'Strider Tester'
      },
      url: 'git@nodev:stridertester/privproject1.git',
      scm: 'git',
      auth: {type: 'ssh'}
    };

    const job = {};

    it('should call back with err as null and the second param as an object with three properties, config, account and a fetch function', function(done) {
      worker.init(dirs, account, config, job, function(err, object) {
        expect(err).to.not.be.ok();
        expect(object.config).to.eql(config);
        expect(object.account).to.eql(account);
        expect(object.fetch).to.be.a('function');
        done();
      });
    });

    it('should check for invalid arguments and operational errors ?');
  });

  //--------------------------------------------------------------------------------------
  describe('fetch', function() {

    it('should call our callback after strider-git/worker fetch is done with its task', function(done) {
      const dest = "";
      const account = {
        accessToken: "abcd"
      };
      const config = {
        auth: {
          type: "https",
          username: "strider"
        }
      };
      const job = {};
      const context = {};

      worker.fetch(dest, account, config, job, context, done);
    });

    it('should check for invalid arguments and operational errors ?');
  });

});


lib/webhooks.js
'use strict';

const debug = require('debug')('strider-gitlab:webhooks');
const gravatar = require('gravatar');
const util = require('util');

module.exports = {
  receiveWebhook: receiveWebhook,
  pushJob: pushJob,
};

function makeJob(project, config) {
  var deploy = false;
  var branch = project.branch(config.branch) || {
    active: true,
    mirror_master: true,
    deploy_on_green: false,
  };

  if (!branch.active) return false;

  if (config.branch !== 'master' && branch.mirror_master) {
    // mirror_master branches don't deploy
    deploy = false;
  } else {
    deploy = config.deploy && branch.deploy_on_green;
  }

  return {
    type: deploy ? 'TEST_AND_DEPLOY' : 'TEST_ONLY',
    trigger: config.trigger,
    project: project.name,
    ref: config.ref,
    plugin_data: config.plugin_data || {},
    user_id: project.creator._id,
    created: new Date(),
  };
}

function startFromCommit(project, payload, send) {
  const config = pushJob(payload);
  if (payload.commits) {
    const lastCommit = payload.commits[payload.commits.length - 1];

    if (lastCommit.message.indexOf('[skip ci]') > -1) {
      return { skipCi: true };
    }
  } else if (payload.after && /^0+$/.test(payload.after)) {
    // Tag push event reaction from tags added or removed
    // When reaction from `remove`, `payload.after` hash is all 0
    return { deleteTag: true };
  }

  var branch = project.branch(config.branch);
  var job;

  if (branch) {
    job = makeJob(project, config);
    if (job) return send(job);
  }

  return false;
}

// returns : {trigger, branch, deploy}
function pushJob(payload) {
  let branchname;
  let commit;
  if (payload.commits) {
    commit = payload.commits[payload.commits.length - 1];

    if (!commit || !commit.author) {
      debug(
        "Unable to determine author from provided commit (may it's a merge commit?). Skipping"
      );
      debug(`Commit data extracted from payload:${util.inspect(commit)}`);
      return null;
    }
  } else if (payload.after) {
    // If data haven't commits array but basic information to start ci exist
    commit = {
      url: `${payload.repository.homepage}/commit/${payload.after}`,
      message: 'tag push',
      timestamp: new Date().toISOString(),
      author: {
        name: payload.user_name || '',
        email: undefined,
        image: '',
      },
    };
  } else {
    return null;
  }

  let trigger;
  let ref;
  var isTagPush = false;

  if (payload.ref.indexOf('refs/heads/') === 0) {
    branchname = payload.ref.substring('refs/heads/'.length);
    ref = {
      branch: branchname,
      id: payload.after,
    };
  } else if (payload.ref.indexOf('refs/tags/') === 0) {
    // `tag` ref is start with refs/tags/
    // Git clone options can set `--branch tag-name` to clone `tag` same as clone `branch`
    isTagPush = true;
    branchname = payload.ref.substring('refs/tags/'.length);
    ref = {
      branch: branchname,
      id: payload.after,
    };
  } else {
    ref = {
      fetch: payload.ref,
    };
  }

  trigger = {
    type: isTagPush ? 'tag push' : 'commit',
    author: {
      name: commit.author.name,
      username: commit.author.username,
      email: commit.author.email,
      image: gravatar.url(commit.author.email, {}, true),
    },
    url: commit.url,
    message: commit.message,
    timestamp: commit.timestamp,
    source: {
      type: 'plugin',
      plugin: 'gitlab',
    },
  };

  return {
    branch: branchname,
    trigger: trigger,
    deploy: true,
    ref: ref,
  };
}

function receiveWebhook(emitter, req, res) {
  var payload = req.body;

  res.sendStatus(204);

  var result = startFromCommit(req.project, payload, sendJob);

  if (result && result.skipCi) {
    debug('Skipping commit due to [skip ci] tag');
  } else if (result && result.deleteTag) {
    debug('Skipping push due to delete tag event');
  } else if (!result) {
    debug('webhook received, but no branches matched or branch is not active');
  }

  function sendJob(job) {
    emitter.emit('job.prepare', job);
    return true;
  }
}


config/config.js
app.controller('GitLabCtrl', [
  '$scope',
  function ($scope) {
    $scope.config = $scope.providerConfig();
    $scope.config.cache = $scope.config.cache || false;

    $scope.addWebhooks = function () {
      $scope.loadingWebhooks = true;
      $.ajax($scope.api_root + 'gitlab/hook', {
        type: 'POST',
        success: function () {
          $scope.loadingWebhooks = false;
          $scope.success('Set gitlab webhooks', true);
        },
        error: function () {
          $scope.loadingWebhooks = false;
          $scope.error('Failed to set gitlab webhooks', true);
        },
      });
    };

    $scope.deleteWebhooks = function () {
      $scope.loadingWebhooks = true;
      $.ajax($scope.api_root + 'gitlab/hook', {
        type: 'DELETE',
        success: function (data) {
          $scope.loadingWebhooks = false;
          $scope.success(data, true);
        },
        error: function () {
          $scope.loadingWebhooks = false;
          $scope.error('Failed to remove gitlab webhooks', true);
        },
      });
    };

    $scope.save = function () {
      this.providerConfig(this.config);
    };
  },
]);


lib/webapp.js
'use strict';

const _ = require('lodash');
const api = require('./api');
const async = require('async');
const debug = require('debug')('strider-gitlab:webapp');
const hostname = process.env.strider_server_name || 'http://localhost:3000';
const util = require('util');
const webhooks = require('./webhooks');

module.exports = {
  // this is the project-level config
  // project.provider.config
  config: {
    url: String,
    owner: String,
    repo: String,
    cache: Boolean,
    pull_requests: { type: String, enum: ['all', 'none', 'whitelist'] },
    whitelist: [
      {
        name: String,
        level: { type: String, enum: ['tester', 'admin'] },
      },
    ],
    // used for the webhook
    secret: String,
    // type: https || ssh
    auth: {},
  },

  listRepos: (config, done) => {
    api.get(config, 'projects', (err, repos) => {
      if (err) return done(err);

      if (!repos) {
        return done(
          new Error(
            'Could not get a list of projects from the GitLab repository. Please check the configuration in Account->GitLab.'
          )
        );
      }

      // Parse and filter only git repos
      done(
        null,
        repos.map(api.parseRepo).filter((repo) => {
          return repo.config.scm === 'git';
        })
      );
    });
  },

  getFile: (filename, ref, account, config, project, done) => {
    const repoId = project.provider.repo_id;
    const branch = encodeURIComponent(ref.branch);
    const encodedFileName = encodeURIComponent(filename);
    const uri = util.format(
      'projects/%d/repository/blobs/%s?filepath=%s',
      repoId,
      branch,
      encodedFileName
    );

    api.get(account, uri, (err, data, res) => {
      if (err) {
        return done(err, null);
      } else {
        done(err, res.text);
      }
    });
  },

  getBranches: (account, config, project, done) => {
    const repoId = project.provider.repo_id;
    const uri = util.format('projects/%d/repository/branches', repoId);

    debug(`Getting URI ${uri}`);
    api.get(account, uri, (err, branches) => {
      debug(
        `We get the following as branches: ${util.inspect(
          branches,
          false,
          10,
          true
        )}`
      );

      done(
        err,
        _.map(branches || [], (branch) => {
          return branch.name;
        })
      );
    });
  },

  // will be namespaced under /:org/:repo/api/gitlab
  routes: (app, context) => {
    app.post('/hook', (req, res) => {
      const url = `${hostname}/${req.project.name}/api/gitlab/webhook`;
      const config = req.accountConfig();

      if (!config.api_key)
        return res.status(400).send('Gitlab account not configured');

      return api.createHooks(
        config,
        req.project.provider.repo_id,
        url,
        (err) => {
          if (err) return res.status(500).send(err.message);

          return res.status(200).send('Webhook registered');
        }
      );
    });

    app.delete('/hook', (req, res) => {
      const url = `${hostname}/${req.project.name}/api/gitlab/webhook`;
      const config = req.accountConfig();

      if (!config.api_key)
        return res.status(400).send('Gitlab account not configured');

      return api.deleteHooks(
        config,
        req.project.provider.repo_id,
        url,
        (err, deleted) => {
          if (err) return res.status(500).send(err.message);

          return res
            .status(200)
            .send(deleted ? 'Webhook removed' : 'No webhook to delete');
        }
      );
    });

    // gitlab should hit this endpoint
    app.anon.post('/webhook', function (req, res) {
      webhooks.receiveWebhook(context.emitter, req, res);
    });
  },

  setupRepo: (account, config, project, done) => {
    const url = `${hostname}/${project.name}/api/gitlab/webhook`;
    const deploy_key_title = util.format('strider-%s', project.name);

    if (!account.api_key)
      return done(new Error('Gitlab account not configured'));

    // Create webhooks and add deploykey
    return async.parallel(
      [
        (callback) => {
          debug('Creating webhooks...');
          api.createHooks(account, project.provider.repo_id, url, callback);
        },
        (callback) => {
          debug('Adding deployment keys...');
          api.addDeployKey(
            account,
            project.provider.repo_id,
            deploy_key_title,
            project.branches[0].pubkey,
            callback
          );
        },
      ],
      (err) => {
        done(err, config);
      }
    );
  },

  teardownRepo: (account, config, project, done) => {
    const url = `${hostname}/${project.name}/api/gitlab/webhook`;
    const deploy_key_title = util.format('strider-%s', project.name);

    if (!account.api_key)
      return done(new Error('Gitlab account not configured'));

    // Remove webhooks and deploykey
    return async.parallel(
      [
        (callback) => {
          api.deleteHooks(account, project.provider.repo_id, url, callback);
        },
        (callback) => {
          api.removeDeployKey(
            account,
            project.provider.repo_id,
            deploy_key_title,
            callback
          );
        },
      ],
      done
    );
  },
};
